LinkedIdeas/CanvasViewController.swift

```diff
@@ -53,6 +70,37 @@ struct DrawableLink: DrawableElement {
   func draw() {
     link.color.set()
     constructArrow()?.bezierPath().fill()
+    drawSelectedRing()
+    drawLinkText()
+  }
+  
+  func drawLinkText() {
+    guard link.stringValue != "" else { return }
+    
+    // background
+    NSColor.white.set()
+    var textSize = link.attributedStringValue.size()
+    let padding: CGFloat = 8.0
+    textSize.width += padding
+    textSize.height += padding
+    
+    let textRect = NSRect(center: link.point, size: textSize)
+    NSRectFill(textRect)
+    
+    // text
+    let bottomLeftTextPoint = link.point.translate(deltaX: -(textSize.width - padding) / 2.0, deltaY: -(textSize.height - padding) / 2.0)
+    let attributedStringValue = NSAttributedString(
+      attributedString: link.attributedStringValue,
+      fontColor: NSColor.gray
+    )
+    attributedStringValue.draw(at: bottomLeftTextPoint)
+  }
+  
+  func drawSelectedRing() {
+    guard link.isSelected else { return }
```

Comment:
I'd give the `return` some newlines. That makes it a bit easier to set breakpoints in the code.